src/function/gds/louvain.cpp

```diff
@@ -0,0 +1,483 @@
+#include <cmath>
+
+#include "binder/binder.h"
+#include "common/exception/runtime.h"
+#include "function/gds/gds_function_collection.h"
+#include "function/gds/gds_utils.h"
+#include "function/gds/gds_utils_mem.h"
+#include "gds_vertex_compute.h"
+#include "graph/graph_mem.h"
+#include "processor/execution_context.h"
+
+using namespace std;
+using namespace kuzu::binder;
+using namespace kuzu::common;
+using namespace kuzu::processor;
+using namespace kuzu::storage;
+using namespace kuzu::graph;
+
+namespace kuzu {
+namespace function {
+
+constexpr uint32_t MAX_LEVELS = 10;
+constexpr uint32_t MAX_ITERATIONS = 10;
+constexpr double THRESHOLD = 1e-6;
+static constexpr offset_t UNASSIGNED_COMM = numeric_limits<offset_t>::max();
+
+struct CommInfo {
+    long size;
+    double degree;
+
+    CommInfo() : size(0), degree(0.0) {}
+};
+
+class CommInfoAtomic {
+public:
+    CommInfoAtomic() : size{0}, degree(0.0) {}
+    CommInfoAtomic(const CommInfoAtomic& other) {
+        size.store(other.getSize());
+        degree.store(other.getDegree());
+    }
+    CommInfoAtomic& operator=(const CommInfoAtomic& other) {
+        if (this != &other) {
+            size.store(other.getSize());
+            degree.store(other.getDegree());
+        }
+        return *this;
+    }
+
+    void addSize(long add) { size.fetch_add(add, memory_order_relaxed); }
+    offset_t getSize() const { return size.load(memory_order_relaxed); }
+    double getDegree() const { return degree.load(memory_order_relaxed); }
+    void addDegree(double add) { degree.fetch_add(add, memory_order_relaxed); }
+
+private:
+    std::atomic<offset_t> size;
+    std::atomic<double> degree;
+};
+
+struct PhaseState {
+    InMemoryGraph graph;
+    vector<double> nodeWeightedDegrees;
+    vector<CommInfo> commInfos;
+    vector<CommInfoAtomic> commInfosNext;
+    vector<offset_t> pastCommunities;
```

Comment:
What's past communities? Is it really needed? If not, let's remove it.

Reply:
Removed

Reply:
Actually `pastCommunities` is needed to store the latest community assignments that have been tested for positive change in modularity. `currComm` stores assignments that is being tested with the new modularity and on failure, the assignments in  `pastCommunities` become the final assignments.

The above is needed because of how the iteration loop computes the modularity. In future, we can again remove this 3rd `acceptedComm` object, but will need a non-trivial refactoring in the modularity computation.

Renamed to `acceptedComm`.